src/main/src/main/java/org/geoserver/security/impl/LayerGroupContainmentCache.java

```diff
@@ -0,0 +1,426 @@
+/* (c) 2016 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.security.impl;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.geoserver.catalog.Catalog;
+import org.geoserver.catalog.CatalogException;
+import org.geoserver.catalog.CatalogInfo;
+import org.geoserver.catalog.LayerGroupInfo;
+import org.geoserver.catalog.LayerInfo;
+import org.geoserver.catalog.PublishedInfo;
+import org.geoserver.catalog.ResourceInfo;
+import org.geoserver.catalog.WorkspaceInfo;
+import org.geoserver.catalog.event.CatalogAddEvent;
+import org.geoserver.catalog.event.CatalogListener;
+import org.geoserver.catalog.event.CatalogModifyEvent;
+import org.geoserver.catalog.event.CatalogPostModifyEvent;
+import org.geoserver.catalog.event.CatalogRemoveEvent;
+
+/**
+ * A cache for layer group containment, it speeds up looking up layer groups containing a particular layer (recursively). * The class is thread safe.
+ *
+ * @author Andrea Aime - GeoSolutions
+ */
+class LayerGroupContainmentCache {
+
+    /**
+     * Builds a concurrent set wrapping a {@link ConcurrentHashMap}
+     */
+    static final Function<? super String, ? extends Set<LayerGroupSummary>> CONCURRENT_SET_BUILDER = k -> Collections
+            .newSetFromMap(new ConcurrentHashMap<LayerGroupSummary, Boolean>());
+
+    /**
+     * Verifies a certain {@link PublishedInfo} is actually a {@link LayerInfo}
+     */
+    static final Predicate<PublishedInfo> IS_LAYER = p -> p.getId() != null
+            && p instanceof LayerInfo && ((LayerInfo) p).getResource() != null;
+
+    /**
+     * Verifies a certain {@link PublishedInfo} is actually a {@link LayerGroupInfo}
+     */
+    static final Predicate<PublishedInfo> IS_GROUP = p -> p.getId() != null
+            && p instanceof LayerGroupInfo;
+
```

Comment:
Nice one, the filter functions (IS_LAYER and IS_GROUP) make the code more readable :)